core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -197,31 +209,38 @@ public long getAllocationDelayTimeoutSetting(Settings settings, Settings indexSe
     }
 
     /**
-     * The time in millisecond until this unassigned shard can be reassigned.
+     * The delay in nanoseconds until this unassigned shard can be reassigned. This value is cached and might be slightly out-of-date.
+     * See also the {@link #updateDelay(long, Settings, Settings)} method.
      */
-    public long getDelayAllocationExpirationIn(long unassignedShardsAllocatedTimestamp, Settings settings, Settings indexSettings) {
-        long delayTimeout = getAllocationDelayTimeoutSetting(settings, indexSettings);
-        if (delayTimeout == 0) {
-            return 0;
-        }
-        long delta = unassignedShardsAllocatedTimestamp - timestamp;
-        // account for time drift, treat it as no timeout
-        if (delta < 0) {
-            return 0;
-        }
-        return delayTimeout - delta;
+    public long getLastComputedDelay() {
+        return lastComputedDelay;
     }
 
+    /**
+     * Updates delay based on current time and index/node settings.
+     * Should only be called from ReplicaShardAllocator.
+     * @return updated delay
+     */
+    public long updateDelay(long nanoTimeNow, Settings settings, Settings indexSettings) {
```

Comment:
Is it possible to refactor this to use the `IndexSettings` class instead of passing two `Settings` instances in?